File: retrocpu_test_framework/src/expand_includes.ts
import fs from "node:fs";
import path from "node:path";

/** リンカ結合すべきソース拡張子（.include 禁止） */
const SOURCE_EXTS = new Set([".asm", ".s", ".mn1610", ".mn1613", ".tms9995"]);

/**
 * 行からコメントを除く（`;` / `//`）。
 * @param line 1 行
 * @returns コメント除去後
 */
function stripLineComment(line: string): string {
  const semi = line.indexOf(";");
  const slash = line.indexOf("//");
  let cut = line.length;
  if (semi >= 0) cut = Math.min(cut, semi);
  if (slash >= 0) cut = Math.min(cut, slash);
  return line.slice(0, cut);
}

/**
 * INCLUDE オペランドからパスを取り出す。
 * @param operandText `.include` のオペランド
 * @returns ファイルパス
 */
function parseIncludeOperand(operandText: string): string {
  const trimmed = operandText.trim();
  if (!trimmed) {
    throw new Error("INCLUDE requires a file path.");
  }
  if (
    (trimmed.startsWith('"') && trimmed.endsWith('"')) ||
    (trimmed.startsWith("'") && trimmed.endsWith("'"))
  ) {
    return trimmed.slice(1, -1).trim();
  }
  return trimmed;
}

/**
 * `.include` を再帰展開する。`.asm` 等のソース取り込みは禁止（リンカで結合する）。
 * 根拠: asm-rules.mdc（`.asm` は include せず `.rel` をリンク）
 * @param sourceText ソース全文
 * @param fromDir `.include` 相対パスの基準ディレクトリ
 * @param includeStack 循環検出用スタック
 * @returns 展開済みソース
 */
export function expandIncludes(
  sourceText: string,
  fromDir: string,
  includeStack: string[] = [],
): string {
  const absDir = path.resolve(fromDir);
  const lines = sourceText.replace(/\r\n/g, "\n").split("\n");
  const out: string[] = [];

  for (let i = 0; i < lines.length; i += 1) {
    const raw = lines[i];
    const body = stripLineComment(raw).trim();
    const m = body.match(/^(?:\.INCLUDE|INCLUDE)\s+(.+)$/i);
    if (!m) {
      out.push(raw);
      continue;
    }

    const includeOperand = parseIncludeOperand(m[1]);
    const includeFile = path.isAbsolute(includeOperand)
      ? includeOperand
      : path.resolve(absDir, includeOperand);
    const absInclude = path.resolve(includeFile);
    const ext = path.extname(absInclude).toLowerCase();

    if (SOURCE_EXTS.has(ext)) {
      throw new Error(
        `Cannot .include assembler source '${includeOperand}' (${absDir}:${i + 1}). ` +
          "Assemble each .asm separately and link with the linker.",
      );
    }
    if (includeStack.includes(absInclude)) {
      throw new Error(
        `Include cycle detected: ${[...includeStack, absInclude].join(" -> ")}`,
      );
    }
    if (!fs.existsSync(absInclude)) {
      throw new Error(
        `Include file not found: ${includeOperand} (${absDir}:${i + 1})`,
      );
    }

    const nested = fs.readFileSync(absInclude, "utf8");
    out.push(
      expandIncludes(nested, path.dirname(absInclude), [
        ...includeStack,
        absInclude,
      ]),
    );
  }

  return out.join("\n");
}

/**
 * ファイル起点で `.include` を展開する。
 * @param entryPath 起点 ASM
 * @returns 展開済みソース
 */
export function expandIncludesFromFile(entryPath: string): string {
  const absPath = path.resolve(entryPath);
  const text = fs.readFileSync(absPath, "utf8");
  return expandIncludes(text, path.dirname(absPath), [absPath]);
}

File: retrocpu_test_framework_rs/src/expand_includes.rs
use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};

use crate::error::FrameworkError;

fn source_exts() -> HashSet<&'static str> {
    HashSet::from(["asm", "s", "mn1613", "tms9995"])
}

fn strip_line_comment(line: &str) -> &str {
    let semi = line.find(';');
    let slash = line.find("//");
    match (semi, slash) {
        (Some(a), Some(b)) => &line[..a.min(b)],
        (Some(a), None) => &line[..a],
        (None, Some(b)) => &line[..b],
        (None, None) => line,
    }
}

fn parse_include_operand(operand_text: &str) -> Result<String, FrameworkError> {
    let trimmed = operand_text.trim();
    if trimmed.is_empty() {
        return Err(FrameworkError::invalid_argument(
            "INCLUDE requires a file path.",
        ));
    }

    if ((trimmed.starts_with('"') && trimmed.ends_with('"'))
        || (trimmed.starts_with('\'') && trimmed.ends_with('\'')))
        && trimmed.len() >= 2
    {
        return Ok(trimmed[1..trimmed.len() - 1].trim().to_string());
    }

    Ok(trimmed.to_string())
}

fn expand_internal(
    source_text: &str,
    from_dir: &Path,
    include_stack: &mut Vec<PathBuf>,
) -> Result<String, FrameworkError> {
    let source_exts = source_exts();
    let abs_dir = from_dir
        .canonicalize()
        .unwrap_or_else(|_| from_dir.to_path_buf());
    let lines = source_text.replace("\r\n", "\n");
    let mut out: Vec<String> = Vec::new();

    for (idx, raw) in lines.split('\n').enumerate() {
        let body = strip_line_comment(raw).trim();
        let upper = body.to_ascii_uppercase();
        let operand = if let Some(rest) = upper.strip_prefix(".INCLUDE ") {
            Some((&body[body.len() - rest.len()..]).trim())
        } else if let Some(rest) = upper.strip_prefix("INCLUDE ") {
            Some((&body[body.len() - rest.len()..]).trim())
        } else {
            None
        };

        let Some(operand_text) = operand else {
            out.push(raw.to_string());
            continue;
        };

        let include_operand = parse_include_operand(operand_text)?;
        let include_file = {
            let p = PathBuf::from(&include_operand);
            if p.is_absolute() {
                p
            } else {
                abs_dir.join(p)
            }
        };

        let abs_include = include_file
            .canonicalize()
            .unwrap_or_else(|_| include_file.clone());

        let ext = abs_include
            .extension()
            .and_then(|s| s.to_str())
            .unwrap_or("")
            .to_ascii_lowercase();
        if source_exts.contains(ext.as_str()) {
            return Err(FrameworkError::invalid_argument(format!(
                "Cannot .include assembler source '{include_operand}' ({}:{}). Assemble each .asm separately and link with the linker.",
                abs_dir.display(),
                idx + 1
            )));
        }

        if include_stack.iter().any(|p| p == &abs_include) {
            let mut chain: Vec<String> = include_stack
                .iter()
                .map(|p| p.to_string_lossy().to_string())
                .collect();
            chain.push(abs_include.to_string_lossy().to_string());
            return Err(FrameworkError::invalid_argument(format!(
                "Include cycle detected: {}",
                chain.join(" -> ")
            )));
        }

        if !include_file.exists() {
            return Err(FrameworkError::invalid_argument(format!(
                "Include file not found: {include_operand} ({}:{})",
                abs_dir.display(),
                idx + 1
            )));
        }

        let nested = fs::read_to_string(&include_file).map_err(|e| {
            FrameworkError::invalid_argument(format!(
                "failed to read include file {}: {e}",
                include_file.display()
            ))
        })?;

        include_stack.push(abs_include.clone());
        let expanded = expand_internal(
            &nested,
            include_file.parent().unwrap_or(abs_dir.as_path()),
            include_stack,
        )?;
        include_stack.pop();

        out.push(expanded);
    }

    Ok(out.join("\n"))
}

pub fn expand_includes(
    source_text: &str,
    from_dir: &Path,
    include_stack: Option<Vec<PathBuf>>,
) -> Result<String, FrameworkError> {
    let mut stack = include_stack.unwrap_or_default();
    expand_internal(source_text, from_dir, &mut stack)
}

pub fn expand_includes_from_file(entry_path: &Path) -> Result<String, FrameworkError> {
    let abs_path = entry_path
        .canonicalize()
        .unwrap_or_else(|_| entry_path.to_path_buf());
    let stack_entry = abs_path.clone();
    let text = fs::read_to_string(&abs_path).map_err(|e| {
        FrameworkError::invalid_argument(format!("failed to read {}: {e}", abs_path.display()))
    })?;
    expand_includes(
        &text,
        abs_path.parent().unwrap_or(Path::new(".")),
        Some(vec![stack_entry]),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::{SystemTime, UNIX_EPOCH};

    fn unique_temp_dir() -> PathBuf {
        let mut p = std::env::temp_dir();
        let nanos = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_nanos();
        p.push(format!("tf-rs-expand-{nanos}"));
        fs::create_dir_all(&p).expect("temp dir should be created");
        p
    }

    #[test]
    fn expands_nested_include() {
        let dir = unique_temp_dir();
        let a = dir.join("a.inc");
        let b = dir.join("b.inc");
        fs::write(&b, "VALUE\n").expect("write b");
        fs::write(&a, ".include \"b.inc\"\n").expect("write a");

        let src = "HEAD\n.include \"a.inc\"\nTAIL\n";
        let out = expand_includes(src, &dir, None).expect("expand should work");
        assert!(out.contains("HEAD"));
        assert!(out.contains("VALUE"));
        assert!(out.contains("TAIL"));

        let _ = fs::remove_file(a);
        let _ = fs::remove_file(b);
        let _ = fs::remove_dir_all(dir);
    }

    #[test]
    fn rejects_asm_include() {
        let dir = unique_temp_dir();
        let e = expand_includes(".include \"foo.asm\"\n", &dir, None)
            .expect_err("asm include should fail");
        assert!(format!("{e}").contains("Cannot .include assembler source"));
        let _ = fs::remove_dir_all(dir);
    }
}
